Point at fn bound that introduced lifetime obligation

estebank opened the pull request:
The last note is new
```
error[E0597]: `c` does not live long enough
  --> $DIR/without-precise-captures-we-are-powerless.rs:19:20
   |
LL | fn simple<'a>(x: &'a i32) {
   |           -- lifetime `'a` defined here
...
LL |     let c = async move || { println!("{}", *x); };
   |         - binding `c` declared here
LL |     outlives::<'a>(c());
   |     ---------------^---
   |     |              |
   |     |              borrowed value does not live long enough
   |     argument requires that `c` is borrowed for `'a`
LL |     outlives::<'a>(call_once(c));
LL | }
   | - `c` dropped here while still borrowed
   |
note: requirement that `c` is borrowed for `'a` introduced here
  --> $DIR/without-precise-captures-we-are-powerless.rs:7:33
   |
LL | fn outlives<'a>(_: impl Sized + 'a) {}
   |                                 ^^
```

When encountering a `ConstraintCategory::Predicate` in a funtion call, point at the `Span` for that `Predicate` to explain where the lifetime obligation originates from.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->


CC rust-lang/rust#55307.

estebank committed:
Point at fn bound that introduced lifetime obligation

```
error[E0597]: `c` does not live long enough
  --> $DIR/without-precise-captures-we-are-powerless.rs:19:20
   |
LL | fn simple<'a>(x: &'a i32) {
   |           -- lifetime `'a` defined here
...
LL |     let c = async move || { println!("{}", *x); };
   |         - binding `c` declared here
LL |     outlives::<'a>(c());
   |     ---------------^---
   |     |              |
   |     |              borrowed value does not live long enough
   |     argument requires that `c` is borrowed for `'a`
LL |     outlives::<'a>(call_once(c));
LL | }
   | - `c` dropped here while still borrowed
   |
note: requirement that `c` is borrowed for `'a` introduced here
  --> $DIR/without-precise-captures-we-are-powerless.rs:7:33
   |
LL | fn outlives<'a>(_: impl Sized + 'a) {}
   |                                 ^^
```

When encountering a `ConstraintCategory::Predicate` in a funtion call, point at the `Span` for that `Predicate` to explain where the lifetime obligation originates from.

estebank committed:
Point at lifetime requirement origin in more cases

lcnr commented on tests/ui/async-await/async-closures/without-precise-captures-we-are-powerless.stderr:
```diff
@@ -29,7 +29,7 @@ LL |     outlives::<'a>(call_once(c));
    | - `c` dropped here while still borrowed
    |
-note: requirement that `c` is borrowed for `'a` introduced here
+note: requirement for `'a` introduced here
```
what is a requirement for `'a`. I feel like the requirement is *for the APIT* here, not `'a`. Is it not possible to keep sth like "requirement that <whatever> outlives <region>" introduced here"?

estebank committed:
reword note

lcnr commented on tests/ui/borrowck/issue-17545.stderr:
```diff
@@ -10,6 +10,9 @@ LL | |     ));
    |        argument requires that borrow lasts for `'a`
+   |
+note: requirement that the value outlives `'a` introduced here
+  --> $SRC_DIR/core/src/ops/function.rs:LL:COL
```
how come this doesn't actually print anything?

estebank replied:
because the span is in `std`, which don't get displayed when the sources aren't found (like when path remap is being used, which is done for all tests).

lcnr reviewed (commented):
r=me after nit

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146011/fe55364329579d361b1ab565728bc033a7dba07e..79982d76a02710575c4a04088d3bb8133bd826a8)*

lcnr commented on compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs:
```diff
@@ -416,6 +416,19 @@ impl<'tcx> BorrowExplanation<'tcx> {
+                        err.span_note(
+                            pred,
+                            format!("requirement that the value outlives `{region_name}` introduced here"),
+                        );
```
actually, idk if that's much easier to read, pls either use an `if let` instead of a loop here or add a comment that we only print this note for the first occurrance

estebank replied:
Changed it to be `if let Some(span) = iter.filter_map(...).next()`.

estebank committed:
Address review comment

estebank committed:
Point at multiple outlives requirements instead of just the first one

```
error[E0716]: temporary value dropped while borrowed
  --> $DIR/multiple-sources-for-outlives-requirement.rs:5:38
   |
LL | fn foo<'b>() {
   |        -- lifetime `'b` defined here
LL |     outlives_indir::<'_, 'b, _>(&mut 1u32);
   |     ---------------------------------^^^^-- temporary value is freed at the end of this statement
   |     |                                |
   |     |                                creates a temporary value which is freed while still in use
   |     argument requires that borrow lasts for `'b`
   |
note: requirements that the value outlives `'b` introduced here
  --> $DIR/multiple-sources-for-outlives-requirement.rs:1:23
   |
LL | fn outlives_indir<'a: 'b, 'b, T: 'a>(_x: T) {}
   |                       ^^         ^^
```

estebank commented on tests/ui/regions/multiple-sources-for-outlives-requirement.stderr:
```diff
@@ -0,0 +1,20 @@
+   |     |                                creates a temporary value which is freed while still in use
+   |     argument requires that borrow lasts for `'b`
+   |
+note: requirements that the value outlives `'b` introduced here
+  --> $DIR/multiple-sources-for-outlives-requirement.rs:1:23
+   |
+LL | fn outlives_indir<'a: 'b, 'b, T: 'a>(_x: T) {}
+   |                       ^^         ^^
```
The last commit is independent and changes it to point at *every* requirement, as previously this error was pointing only at the second span, which even though is correct, it isn't very understandable to someone who doesn't have a good grasp of Rust yet. What do you think @lcnr? I can drop the last commit and make it point just at the `'a` here.

lcnr replied:
hmm, I agree that this seems better than just pointing at one of them🤔

lcnr commented:
@bors r+ rollup
